File: InvoiceApi.NET/Models/Payment.cs
namespace InvoiceApi.NET.Models;

public class Payment
{
    public string PaymentId { get; set; } = string.Empty;
    public string PayerName { get; set; } = string.Empty;
    public decimal Amount { get; set; } 
    public string? Reference { get; set; }
    public DateOnly PaymentDate { get; set; }

    // if matched, the following is recorded
    public string? MatchedInvoiceId { get; set; }
    public string? MatchRule { get; set; }
    public double? Confidence { get; set; }

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
}

File: cash_allocator_mobile/lib/models/payment.dart
class Payment {
  final String paymentId;
  final String payerName;
  final double amount;
  final String? reference;
  final String paymentDate;

  final String? matchedInvoiceId;
  final String? matchRule;
  final double? confidence;

  final String createdAt;

  Payment({
    required this.paymentId,
    required this.payerName,
    required this.amount,
    required this.reference,
    required this.paymentDate,
    required this.matchedInvoiceId,
    required this.matchRule,
    required this.confidence,
    required this.createdAt,
  });

  factory Payment.fromJson(Map<String, dynamic> json) {
    return Payment(
      paymentId: json['paymentId'] as String,
      payerName: json['payerName'] as String,
      amount: (json['amount'] as num).toDouble(),
      reference: json['reference'] as String?,
      paymentDate: json['paymentDate'] as String,
      matchedInvoiceId: json['matchedInvoiceId'] as String?,
      matchRule: json['matchRule'] as String?,
      confidence: (json['confidence'] as num).toDouble(),
      createdAt: json['createdAt'] as String,
    );
  }

  bool get needsReview => matchRule == 'fuzzy_match' && matchedInvoiceId != null;
  bool get isAutoMatched => matchRule == 'reference_match' || matchRule == 'exact_amount_name';
  bool get isUnmatched => matchRule == null;
}